### ISSUE TITLE
关于adam优化器中的weight_decay

### ISSUE THREAD

**Issue by Light-lyy** [THUwangcy/DirectAU#2]
请问在adam优化器的参数中设定weight_decay（您在recbole中的实现方式），与在计算loss时加入向量的L2正则化损失（许多其他论文的代码实现）相比，有什么不同吗？

也就是说不设定adam优化器参数中的weight_decay，而是在calculate_loss(self, user, item)中这样写
loss = align + self.gamma * uniform **+ weight_decay\*(user_embedding[user].norm(2).pow(2)+item_embedding[item].norm(2).pow(2))**
带来的不同效果是什么呢？

我先是像您那样设定adam优化器的weight_decay参数，得到了不错的结果。但用后一种方法我得到的结果远不如前一种方法。我不太理解其中缘由（以前我看过的论文代码都是用后一种方法，这是我第一次见到像您这样用到了优化器中的weight_decay参数），希望能得到您的解答

**Comment by THUwangcy** [comment]
weight_decay 和 L2 正则只有在 SGD 优化器的情况下才是等价的，Adam 优化器就会存在差异，最优的正则项系数也会不一样，之前有工作也探究过 weight_decay 和 L2 的区别（ https://arxiv.org/pdf/1711.05101.pdf ）。我理解实际使用中两种方式其实均可，只是要分别调参。
